### ISSUE TITLE
More retail (SA) - compliance check

### ISSUE THREAD

**Issue by sahil-ondc** [ONDC-Official/v1.2.0-logs#756]
### on_search_full_catalog_refresh

- /message/catalog/bpp/descriptor must have required property 'tags'
- /message/catalog/bpp/providers/locations/time/schedule/holidays cannot be past dates '2023-12-31'
- /message/catalog/bpp/providers/items/quantity/unitized/unit should be as per API contract
- 'manufacturer_or_packer_name', 'manufacturer_or_packer_address', 'common_or_generic_name_of_commodity', 'net_quantity_or_measure_of_commodity_in_pkg', 'month_year_of_manufacture_packing_import' should not be provided as '/'

### on_init

- /message/order should not have property 'provider_location' (Deprecated)
- 'settlement_bank_account_no', 'settlement_ifsc_code', 'branch_name' should not have placeholder values

### on_confirm

- provider_tax_number (pan number of merchant) should be provided in /tags/bpp_terms

### on_select (out-of-stock)

- error/message should be exactly as per API contract
  `error.message in the response should have the list of corresponding item ids referenced in stringified object notation, e.g. if items "I1" is out of stock, error.message would be encoded as: "[{\"item_id\":\"I1\",\"error\":\"40002\"}]"`

### on_cancel

- value should be string in quote_trail
- context/action should be on_cancel

### on_update

- 'settlement_bank_account_no', 'settlement_ifsc_code', 'branch_name' should not have placeholder values

## Note

- packing charge to be included only if not 0
- tax breakup to be included only if tax is not 0
- misc to be included only if not 0;

**Comment by chinmaya1920** [comment]
Hai @sahil-ondc 

Can you check this [pull request](https://github.com/ONDC-Official/v1.2.0-logs/pull/907)?

And confirm on the points please.

**Comment by sahil-ondc** [comment]
## Flow 1

### on_search_full_catalog_refresh

- 'accept_bap_terms' should not be provided as its not part of phase 1
- /message/catalog/bpp/fulfillments/id and /message/catalog/bpp/providers/fulfillments/id should be mapped
- /message/catalog/bpp/providers/descriptor short_desc and long_desc should not have placeholder values
- radius provided in /message/catalog/bpp/providers/locations/circle and /message/catalog/bpp/providers/tags/serviceability should be same
- /message/catalog/bpp/providers/items should not have value '@ondc/org/mandatory_reqs_veggies_fruits' (deprecated)
- either one of '@ondc/org/statutory_reqs_packaged_commodities' and '@ondc/org/statutory_reqs_prepackaged_food' should be provided as per the sheet
  `https://docs.google.com/spreadsheets/d/1APAvavF_BNbTA89benAlGtv0GuFvpn2b6XXi4lSdTTw/edit#gid=0`
- invalid value provided '/' for 'manufacturer_or_packer_address', 'common_or_generic_name_of_commodity', 'net_quantity_or_measure_of_commodity_in_pkg', 'month_year_of_manufacture_packing_import', 'nutritional_info', 'additives_info', 'brand_owner_FSSAI_license_no', 'other_FSSAI_license_no', 'importer_FSSAI_license_no'

### on_search_inc

- /message/catalog/bpp/providers should not contain properties 'fulfillments', 'descriptor', 'ttl', 'locations'

## Flow 2

### on_select

- /message/order/quote/breakup/item/quantity available and maximum should be same as provided in on_search

### on_init

- 'settlement_bank_account_no', 'settlement_ifsc_code', 'branch_name' should not have placeholder values

### on_confirm

- 'provider_tax_number' should be provided in bpp_terms in /message/order/tags

### on_status

- missing order state (packed, agent-assigned, out-for-delivery)

## Flow 3

### on_select (out-of-stock)

- /error/message should have unique item_id (duplicate item_id: 102074644)

## Flow 4

### on_cancel

- /message/order/fulfillment/tags/cancel_request/initiated_by should be bap_id
- 'settlement_bank_account_no', 'settlement_ifsc_code', 'branch_name' should not have placeholder values

## Flow 5

### on_cancel

- invalid value '2.0' provided in /message/order/items/quantity/count
- quote_trail should be provided for entire cancelled items
- /message/order/fulfillment/return/start/time/timestamp should be provided in return state picked

## Flow 6

- Flow should be provided as per the sheet
  `https://docs.google.com/spreadsheets/d/1JZV6ZQzXcHUsOwegGtArX3DdIXYIy3gxkhQ00q7kICc/edit#gid=1367601795`

## Note

- packing charge to be included only if not 0
- tax breakup to be included only if tax is not 0
- misc to be included only if not 0;

@chinmaya1920

**Comment by rahul-repository** [comment]
@sahil-ondc are the **'settlement_bank_account_no', 'settlement_ifsc_code', 'branch_name'**  needed? I see in the ondc contract 1.2.0 settlement details are optional field. We just added it in the response because the log validation api were failing without it. Please confirm.

**Comment by rahul-repository** [comment]
@sahil-ondc 

Flow 5 - /message/order/fulfillment/return/start/time/timestamp should be provided in return state picked

In this point we are not able to find return key inside fulfillment. We are providing timestamp for the fullfillment RTO-Initiated. 
Can you just guide us through this where exactly we are not sending the timestamp

**Comment by bluecypher** [comment]
@rahul-repository, these are the settlement details that are required to be provided in order to receive the settlement amount. 

For Flow 6, liquidation flow is mandatory, reverse QC flow is optional for grocery category

@sahil-ondc, please help with the issue in Flow 5

**Comment by rahul-repository** [comment]
Hi @bluecypher according to the ONDC contract 1.2.0 we have the following scenarios listed below for return with liquidation.

Scenario A - Return liquidated:
- Return_Initiated -> Return_Liquidated (refund triggered here)

Scenario B - Return rejected:
- Return_Initiated -> Return_Rejected

In our case we are handling the Scenario B.  So is it fine we submit only scenario B? or is it mandatory to submit both the scenarios?

**Comment by sahil-ondc** [comment]
## Flow 2

### on_status

- Created At timestamp for /on_status_pending should be equal to context timestamp at confirm

## Flow 4

### on_select

- context/timestamp difference between /on_select and /select should be less than 5 sec

### on_cancel

- /message/order/provider/id must be string
- /message/order/fulfillments/0/start/location/id must be string
- /message/order/fulfillments/0/end must have required property 'person'
- /message/order/payment must have required property 'params'
- /message/order/payment must have required property 'status'
- /message/order/payment must have required property 'type'
- /message/order/payment must have required property 'collected_by'
- Timestamp for /on_confirm api cannot be greater than or equal to /on_cancel api
- The count of cancellation fulfillmentsn is not equal to the count of forward fulfillments or invalid fulfillment id.
- Key length mismatch for root when compared with init billing object
- Pre Cancel is mandatory for on_cancel

## Flow 5

### select

- City code mismatch in /on_search_full_catalog_refresh and /select
- Invalid item found in /select id: 101205877
- Invalid item found in /select id: 100066264
- Invalid item found in /select id: 101815423
- provider with provider.id: 1594 does not exist in on_search
- Warning: Missed checks for provider as provider with ID: 1594 does not exist on on_search API"

### on_select

- /fulfillments/@ondc/org/TAT (O2D) in /on_select can't be less than @ondc/org/time_ship (O2S) in /on_search

### on_confirm

- Invalid store name inside fulfillments in /on_confirm

### on_status_picked

- fulfillments/state should be Order-picked-up for /on_status_Order-picked-up

### on_status_out_for_delivery

- fulfillments/state Should be out-for-delivery for /on_status_Out-for-delivery

### on_cancel

- /message/order/id Should be alphanumeric upto 32 letters max
- /message/order/provider/id must be string
- /message/order/fulfillments/0/start/location must have required property 'id'
- /message/order/fulfillments/0/end must have required property 'person'
- /message/order/fulfillments/0/tags/0/list/0/value must be string
- /message/order/payment must have required property 'params'
- /message/order/payment must have required property 'status'
- /message/order/payment must have required property 'type'
- /message/order/payment must have required property 'collected_by'
- City code mismatch in /search and /on_cancel
- Key length mismatch for root when compared with init billing object
- order.created_at timestamp should match context.timestamp
- Pre Cancel is mandatory for on_cancel

## Flow6-a

### on_confirm

- np_type is missing in message/order/tags/bpp_terms for ON_CONFIRM

### on_update

- /message/order/state must be equal to one of the allowed values (Created,Accepted,Cancelled,Completed,Delivered)
- /message/order/billing must have required property 'email'
- /message/order/fulfillments/0/state/descriptor/code must be equal to one of the allowed values (Cancelled,Pending,Order-delivered,Return_Initiated,Return_Pick_Failed,Return_Approved,Return_Picked,Return_Delivered,Liquidated,Return_Rejected)
- /message/order/fulfillments/0/start/location/id must be string
- /message/order/fulfillments/0/start/contact must have required property 'email'
- Item quote breakup prices for /on_update shoulb be equal to the net price.

### update

- Missing required data of : update

## Flow 6-c

### on_init

- Payment details are missing/on_init

### on_update_interim

- /message/order/items/2/quantity/count must be string
- /message/order/billing must have required property 'email'
- /message/order/fulfillments/0/start/location/id must be string
- /message/order/fulfillments/0/start/contact must have required property 'email'
- ttl = PT30S as per the API Contract

### on_update_liquidated

- /message/order/items/2/quantity/count must be string
- /message/order/billing must have required property 'email'
- /message/order/fulfillments/0/start/location/id must be string
- /message/order/fulfillments/0/start/contact must have required property 'email'

**Comment by sahil-ondc** [comment]
## Flow 2

### on_status

- Created At timestamp for /on_status_pending should be equal to context timestamp at confirm

## Flow 4

### on_select

- context/timestamp difference between /on_select and /select should be less than 5 sec

### on_cancel

- /message/order/provider/id must be string
- /message/order/fulfillments/0/start/location/id must be string
- /message/order/fulfillments/0/end must have required property 'person'
- /message/order/payment must have required property 'params'
- /message/order/payment must have required property 'status'
- /message/order/payment must have required property 'type'
- /message/order/payment must have required property 'collected_by'
- Timestamp for /on_confirm api cannot be greater than or equal to /on_cancel api
- The count of cancellation fulfillmentsn is not equal to the count of forward fulfillments or invalid fulfillment id.
- Key length mismatch for root when compared with init billing object
- Pre Cancel is mandatory for on_cancel

## Flow 5

### select

- City code mismatch in /on_search_full_catalog_refresh and /select
- Invalid item found in /select id: 101205877
- Invalid item found in /select id: 100066264
- Invalid item found in /select id: 101815423
- provider with provider.id: 1594 does not exist in on_search
- Warning: Missed checks for provider as provider with ID: 1594 does not exist on on_search API"

### on_select

- /fulfillments/@ondc/org/TAT (O2D) in /on_select can't be less than @ondc/org/time_ship (O2S) in /on_search

### on_confirm

- Invalid store name inside fulfillments in /on_confirm

### on_status_picked

- fulfillments/state should be Order-picked-up for /on_status_Order-picked-up

### on_status_out_for_delivery

- fulfillments/state Should be out-for-delivery for /on_status_Out-for-delivery

### on_cancel

- /message/order/id Should be alphanumeric upto 32 letters max
- /message/order/provider/id must be string
- /message/order/fulfillments/0/start/location must have required property 'id'
- /message/order/fulfillments/0/end must have required property 'person'
- /message/order/fulfillments/0/tags/0/list/0/value must be string
- /message/order/payment must have required property 'params'
- /message/order/payment must have required property 'status'
- /message/order/payment must have required property 'type'
- /message/order/payment must have required property 'collected_by'
- City code mismatch in /search and /on_cancel
- Key length mismatch for root when compared with init billing object
- order.created_at timestamp should match context.timestamp
- Pre Cancel is mandatory for on_cancel

## Flow6-a

### on_confirm

- np_type is missing in message/order/tags/bpp_terms for ON_CONFIRM

### on_update

- /message/order/state must be equal to one of the allowed values (Created,Accepted,Cancelled,Completed,Delivered)
- /message/order/billing must have required property 'email'
- /message/order/fulfillments/0/state/descriptor/code must be equal to one of the allowed values (Cancelled,Pending,Order-delivered,Return_Initiated,Return_Pick_Failed,Return_Approved,Return_Picked,Return_Delivered,Liquidated,Return_Rejected)
- /message/order/fulfillments/0/start/location/id must be string
- /message/order/fulfillments/0/start/contact must have required property 'email'
- Item quote breakup prices for /on_update shoulb be equal to the net price.

### update

- Missing required data of : update

## Flow 6-c

### on_init

- Payment details are missing/on_init

### on_update_interim

- /message/order/items/2/quantity/count must be string
- /message/order/billing must have required property 'email'
- /message/order/fulfillments/0/start/location/id must be string
- /message/order/fulfillments/0/start/contact must have required property 'email'
- ttl = PT30S as per the API Contract

### on_update_liquidated

- /message/order/items/2/quantity/count must be string
- /message/order/billing must have required property 'email'
- /message/order/fulfillments/0/start/location/id must be string
- /message/order/fulfillments/0/start/contact must have required property 'email'
@chinmaya1920

**Comment by sandeepshahi** [comment]
## Flow 1

**/on_search**

- remove accept_bap_terms object as mentioned in the previous feedback
- collect_payment object is optional
- store's order related timing is not provided; if timings are same for Delivery and Order, use enum "All"
- aren't variants supported?
- "@ondc/org/mandatory_reqs_veggies_fruits" is not required; unit measurement is already covered in quantity/unitized/measure
- valid values of all the statutory reqs must be provided in "@ondc/org/statutory_reqs_packaged_commodities"

## Flow 2

**/on_init**

- cancellation_terms need not be provided (not yet enabled on the network)

**/on_confirm**

- why is delivery TAT changing in /on_select and /on_confirm
- PCC code in /start/instructions should be generated when the order is "Packed"

**/on_status**

- Order-picked-up

  - actual order pickup timestamp is not captured correctly; is even before the order was packed

- Out-for-delivery

  - actual delivery timestamp can't be provided when the order is just OFD

## Flow 3

- similar issues as above

**/on_select**

- out of stock information must also be provided in incremental refresh

## Flow 4

**/on_cancel**

- items mapped with forward shipment must also be provided in /items with quantity/count=0
- Why are two new fulfillment created to map the cancelled items instead of handling it through one fulfillment?

## Flow 5

**/on_cancel**

- actual pickup timestamp missing in forward shipment fulfillment (id = "F1")
- line items with price value = 0 need not be provided in quote_trail

**/on_status**

- several inconsistencies in the order object; check every attributes and objects inside the order object and compare with RTO cancellation object


@chinmaya1920, check for these issues too

**Comment by sandeepshahi** [comment]
## Flow 1

**/on_search**

- np_type is required property in /bpp/descriptor/tags
- aren't variants supported?
- valid values of all the statutory reqs must be provided in "@ondc/org/statutory_reqs_packaged_commodities"
- locations/street should not contain entire address information;

**/on_search (inc)**

- store timings need not be provided again in incremental pull

## Flow 2

**/on_select**

- available count should be either 0 or 99 to specify the availability of stock; no need to provide actual inventory count

**/on_init**

- add provider_tax_number in /tags/bpp_terms
- beneficiary_name must be provided in "@ondc/org/settlement_details"

**/on_status**

- order state can't revert from 'Accepted' in /on_confirm to 'Created' in /on_status later
- gps lat/long must have a minimum decimal precision of 6 digits
- estimated pickup time range can't be before context/timestamp in "Pending" state
- routing information in /fulfillments/tags must be provided on or before "Agent-assigned" state
- value of "@ondc/org/TAT" can't be changed in /fulfillments; check all the interim calls
- invoice should only be provided once the order is picked; invoice link must be accessible
- billing/name should not change

- Order-picked-up

  - actual order pickup timestamp is not captured correctly; is even before the order was packed

## Flow 3

- similar issues as above

**/on_select**

- out of stock information must also be provided in incremental refresh

## Flow 4

**/on_cancel**

- order/created_at should not change; must remain the same as in /confirm; This is applicable for all the APIs

## Flow 5

**/on_status**

- estimated pickup and delivery time ranges are not accurate; should map with the estimated TAT; check this issue in other flows too

**/on_cancel**

- updated_at in precancel_state is not captured correctly

**/on_status (RTO)**

- RTO must be delivered/disposed through /on_status; unsolicited call is missing

## Flow 6

- 6a,6b and 6c must be part of the same order

**/on_update (part cancel)**

- line items with price=0 need not be provided in quote_trail; tax is not even a part of quote/breakup
- adjusted amount of item price should be -ve in quote_trail

**/update (payment)**

- refund in /settlement_amount should be a +ve value; SNP must respond with a NACK

**/on_update (Return Initiation)**

- all quantities of item: "100012084" can't be mapped to both forward shipment "F1" and reverse shipment "65fabe9d5a59e578d7ef781a"
- items must be mapped only once the order is liquidated
- initiated_by is missing in /tags/return_request
- line items with price=0 need not be provided in quote_trail; tax is not even a part of quote/breakup

**/on_update (Liquidated)**

- why is the timestamp difference between return initiation and liquidated is large?


@chinmaya1920

**Comment by sandeepshahi** [comment]
## Flow 1

**/on_search**

- variants must be supported

**/on_search (inc)**

- empty tags attribute need not be provided

## Flow 5

**/on_cancel (RTO)**

- updated_at in precancel_state is not captured correctly; should be of Out-for-delivery state
- cancellation reason id mismatch in /cancellation object and /fulfillments/tags/cancel_request/reason_id
- start and end location must be provided in RTO cancellation fulfillment

**/on_status (RTO)**

- RTO-Disposed timestamp must be provided in /fulfillments/end/time/timestamp for RTO fulfillment

@chinmaya1920, please resubmit logs for Flow 5

**Comment by sandeepshahi** [comment]
## Flow 5

**/on_confirm**

- on_confirm should not take longer than 5 sec to be responded; BNPs might timeout and initiate a retry; Also, /confirm retries must be supported as mentioned in the contract

**/on_cancel (RTO)**

- invalid time ranges for RTO shipment; can't be same as forward shipment ranges

**/on_status (RTO)**

- no need to provide invoice once the order is cancelled or RTO'd


@chinmaya1920

**Comment by rahul-repository** [comment]
@sandeepshahi 

**/on_cancel (RTO)**
As the time ranges for RTO shipment is not mentioned in ONDC API Contract 1.2.0. we will not be adding it.
I hope this works.
Please let us know if its mandatory.

**Comment by sandeepshahi** [comment]
> @sandeepshahi
> 
> **/on_cancel (RTO)** As the time ranges for RTO shipment is not mentioned in ONDC API Contract 1.2.0. we will not be adding it. I hope this works. Please let us know if its mandatory.

Yes, this works.

**Comment by rahul-repository** [comment]
@sandeepshahi 

Considering that the variants represent a significant business change and will affect various modules, our aim is to address this within the next 6-7 months.

**Comment by rahul-repository** [comment]
@sandeepshahi 

Considering that the variants represent a significant business change and will affect various modules, our aim is to address this within the next 6-7 months.
